var/spack/repos/builtin/packages/magma/package.py

```diff
@@ -22,28 +22,37 @@
 # License along with this program; if not, write to the Free Software
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 ##############################################################################
+
 from spack import *
 
 
 class Magma(CMakePackage):
-    """The MAGMA project aims to develop a dense linear algebra library
-    similar to LAPACK but for heterogeneous/hybrid architectures,
-    starting with current "Multicore+GPU" systems.
+    """The MAGMA project aims to develop a dense linear algebra library similar to
+       LAPACK but for heterogeneous/hybrid architectures, starting with current
+       "Multicore+GPU" systems.
     """
 
     homepage = "http://icl.cs.utk.edu/magma/"
     url = "http://icl.cs.utk.edu/projectsfiles/magma/downloads/magma-2.2.0.tar.gz"
 
+    version('2.3.0', '9aaf85a338d3a17303e0c69f86f0ec52')
     version('2.2.0', '6c1ebf4cdf63eb302ff6258ff8c49217')
 
     variant('fortran', default=True,
             description='Enable Fortran bindings support')
+    variant('shared', default=True,
+            description='Enable shared library')
 
+    depends_on('blas')
     depends_on('lapack')
-    depends_on('cuda@9.0:', when='%gcc@6.0:6.9.9')
-    depends_on('cuda@8.0:', when='%gcc@5.0:')
+    depends_on('cuda')
+
+    conflicts('%gcc@6:', when='^cuda@:8')
+    conflicts('%gcc@7:', when='^cuda@:9')
```

Comment:
@tgamblin et al: do these need to be conflicts rather than dependencies? As a quick hack/test, I got a successful install by deleting the conflict marker and explicitly using gcc-6.

```
@@ -69,6 +69,9 @@ class Magma(CMakePackage):
             (spec['lapack'].libs + spec['blas'].libs).joined(';')
         ])
 
+        options += ['-DCMAKE_C_COMPILER=gcc-6',
+                    '-DCMAKE_CXX_COMPILER=g++-6']
+
         options += ['-DBUILD_SHARED_LIBS=%s' %
                     ('ON' if ('+shared' in spec) else 'OFF')]
 
```

Reply:
So, I wish we could do it the more natural way you propose, but compilers aren’t proper dependencies (yet). They’re handled differently so that we can mix compilers in the DAG, so these actually have to be conflicts for now. If you depend on the gcc package, it’ll build gcc for you, which is likely *not* what you want.

More details on future plans here: https://fosdem.org/2018/schedule/event/spack/

Reply:
As I recently learned, a better solution for adding a `cuda` variant is to use the `CudaPackage` class - that will bring in even more known conflicts about cuda as well as the `cuda_arch` variant.

Reply:
Actually, for packages that always depend on MAGMA (so they do not need `cuda` variant) using the `CudaPackage` is not quite what we want - we do not want the `cuda` variant just the `cuda_arch` and some of the other stuff like the conflicts.

Reply:
> we do not want the cuda variant just the cuda_arch and some of the other stuff like the conflicts.

true. I wish there is a way to refactor it into two classes and say keep all conflicts in one together with `cuda_arch` and make it optional dependency via the second guy, but that's certainly outside of the scope of this PR